ComponentKit/Core/ComponentTree/CKComponentGroupTreeNode.mm

```diff
@@ -0,0 +1,66 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "CKComponentGroupTreeNode.h"
+
+#import <ComponentKit/CKComponent.h>
+#import <ComponentKit/CKComponentInternal.h>
+#import "CKThreadLocalComponentScope.h"
+#import <ComponentKit/CKInternalHelpers.h>
+
+#include <tuple>
+
+#import "CKMutex.h"
+
+struct CKComponentTreeNodeComparator {
+  bool operator() (const std::tuple<Class, NSUInteger> &lhs, const std::tuple<Class, NSUInteger> &rhs) const
+  {
+    return std::get<0>(lhs) == std::get<0>(rhs) && std::get<1>(lhs) == std::get<1>(rhs);
+  }
+};
+
+struct CKComponentTreeNodeHasher {
+  std::size_t operator() (const std::tuple<Class, NSUInteger> &n) const
+  {
+    return [std::get<0>(n) hash] ^ std::get<1>(n);
+  }
+};
+
+typedef std::unordered_map<std::tuple<Class, NSUInteger>, CKComponentTreeNode *, CKComponentTreeNodeHasher, CKComponentTreeNodeComparator> CKScopeNodeMap;
+
+@implementation CKComponentGroupTreeNode
+{
+  CKScopeNodeMap _children;
```

Comment:
A note on why the mutation of this map is safe is needed somewhere in this file.